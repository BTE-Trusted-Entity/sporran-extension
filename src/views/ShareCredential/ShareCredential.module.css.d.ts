declare namespace ShareCredentialModuleCssNamespace {
  export interface IShareCredentialModuleCss {
    buttonsLine: string;
    cancel: string;
    checkbox: string;
    container: string;
    credentials: string;
    heading: string;
    label: string;
    password: string;
    passwordLine: string;
    subline: string;
    submit: string;
    valid: string;
  }
}

declare const ShareCredentialModuleCssModule: ShareCredentialModuleCssNamespace.IShareCredentialModuleCss & {
  /** WARNING: Only available when `css-loader` is used without `style-loader` or `mini-css-extract-plugin` */
  locals: ShareCredentialModuleCssNamespace.IShareCredentialModuleCss;
};

export = ShareCredentialModuleCssModule;
