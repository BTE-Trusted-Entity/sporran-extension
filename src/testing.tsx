export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
