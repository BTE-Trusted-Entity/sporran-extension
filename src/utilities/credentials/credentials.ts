import { useContext, useEffect, useMemo } from 'react';
import { isEqual, omit, pick, pull, reject, without } from 'lodash-es';
import {
  Attestation,
  ConfigService,
  Credential,
  Did,
  DidUri,
  ICredential,
  KiltPublishedCredentialCollectionV1,
} from '@kiltprotocol/sdk-js';
import { mutate } from 'swr';

import { storage } from '../storage/storage';
import { parseDidUri } from '../did/did';
import { jsonToBase64 } from '../base64/base64';

import { CredentialsContext } from './CredentialsContext';

type AttestationStatus = 'pending' | 'attested' | 'revoked' | 'invalid';

export interface SporranCredential {
  credential: ICredential;
  name: string;
  cTypeTitle: string;
  attester: string;
  status: AttestationStatus;
  isDownloaded?: boolean;
}

export interface SharedCredential {
  sporranCredential: SporranCredential;
  sharedContents: string[];
}

function toKey(hash: string): string {
  return `credential:${hash}`;
}

export const LIST_KEY = toKey('list');

export async function getList(): Promise<string[]> {
  return (await storage.get(LIST_KEY))[LIST_KEY] || [];
}

async function saveList(list: string[]): Promise<void> {
  await storage.set({ [LIST_KEY]: list });
  await mutate(['getCredentials', LIST_KEY]);
}

export async function saveCredential(
  credential: SporranCredential,
): Promise<void> {
  const key = toKey(credential.credential.rootHash);
  await storage.set({ [key]: credential });
  await mutate(key);

  const list = await getList();
  if (list.includes(key)) {
    await mutate(['getCredentials', LIST_KEY]);
    return;
  }
  list.push(key);
  await saveList(list);
}

interface LegacySporranCredential
  extends Omit<SporranCredential, 'credential'> {
  request: ICredential & { claimerSignature?: unknown };
}

function isLegacySporranCredential(
  input: unknown,
): input is LegacySporranCredential {
  return typeof input === 'object' && input !== null && 'request' in input;
}

// SDK <0.29 had claimerSignature in ICredential
interface LegacyICredential extends ICredential {
  claimerSignature?: unknown;
}

function isLegacyICredential(input: unknown): input is LegacyICredential {
  return (
    typeof input === 'object' && input !== null && 'claimerSignature' in input
  );
}

export function updateLegacyCredential(sporranCredential: SporranCredential) {
  const modernCredential = !isLegacySporranCredential(sporranCredential)
    ? sporranCredential
    : {
        ...omit(sporranCredential, 'request'),
        credential: sporranCredential.request,
      };

  const { credential } = modernCredential;
  if (isLegacyICredential(credential)) {
    delete credential.claimerSignature;
  }

  return modernCredential;
}

async function updateLegacyStorage(keys: string[]) {
  const result: Record<string, SporranCredential> = await storage.get(keys);

  for (const [key, sporranCredential] of Object.entries(result)) {
    const modernCredential = updateLegacyCredential(sporranCredential);
    const updateStorage = !isEqual(sporranCredential, modernCredential);
    if (updateStorage) {
      await storage.set({ [key]: modernCredential });
    }
  }
}

export async function getCredentials(
  keys: string[],
): Promise<SporranCredential[]> {
  await updateLegacyStorage(keys);

  const result = await storage.get(keys);
  const credentials = pick(result, keys);
  return Object.values(credentials);
}

export async function deleteCredential(
  credential: SporranCredential,
): Promise<void> {
  const key = toKey(credential.credential.rootHash);
  await storage.remove(key);

  const list = await getList();
  pull(list, key);

  await saveList(list);
}

export function useCredentials(): SporranCredential[] | undefined {
  return useContext(CredentialsContext);
}

export function isUnusableCredential({ status }: SporranCredential) {
  return ['revoked', 'invalid'].includes(status);
}

export function useIdentityCredentials(
  did: DidUri | undefined,
  onlyUsable = true,
): SporranCredential[] | undefined {
  const all = useCredentials();

  return useMemo(() => {
    if (!all) {
      // storage data pending
      return undefined;
    }
    if (!did) {
      // could be a legacy identity without DID
      return [];
    }
    const usable = reject(all, isUnusableCredential);
    const unusable = all.filter(isUnusableCredential);
    const sorted = [...unusable, ...usable]; // will be displayed in the reverse order
    const filtered = onlyUsable ? usable : sorted;

    const { fullDid } = parseDidUri(did);
    return filtered.filter((credential) =>
      Did.isSameSubject(credential.credential.claim.owner, fullDid),
    );
  }, [all, did, onlyUsable]);
}

async function syncCredentialStatusWithChain(credential: SporranCredential) {
  const initialStatus = credential.status;
  const claimHash = credential.credential.rootHash;

  // revoked and invalid statuses are permanent
  if (['revoked', 'invalid'].includes(initialStatus)) {
    return;
  }

  try {
    const api = ConfigService.get('api');
    const chain = await api.query.attestation.attestations(claimHash);

    const attestation = Attestation.fromChain(chain, claimHash);

    const hasBecomeRevoked = attestation.revoked;
    const hasBecomeAttested = initialStatus === 'pending' && !hasBecomeRevoked;

    if (hasBecomeRevoked) {
      await saveCredential({ ...credential, status: 'revoked' });
    }
    if (hasBecomeAttested) {
      await saveCredential({ ...credential, status: 'attested' });
    }
  } catch {
    // not on chain yet, ignore
  }
}

export function usePendingCredentialCheck(
  credential: SporranCredential | undefined,
): void {
  useEffect(() => {
    if (!credential || credential.status !== 'pending') {
      return;
    }
    (async () => {
      await syncCredentialStatusWithChain(credential);
    })();
  }, [credential]);
}

interface CredentialDownload {
  name: string;
  url: string;
}

export function getCredentialDownload(
  credential: SporranCredential,
): CredentialDownload {
  const name = `${credential.name}-${credential.cTypeTitle}.json`;

  const blob = jsonToBase64(credential);
  const url = `data:text/json;base64,${blob}`;

  return { name, url };
}

export function getUnsignedPresentationDownload(
  sporranCredential: SporranCredential,
  properties: string[],
): CredentialDownload {
  const name = `${sporranCredential.name}-${sporranCredential.cTypeTitle}-presentation.json`;

  const { credential } = sporranCredential;
  const allProperties = Object.keys(credential.claim.contents);
  const needRemoving = without(allProperties, ...properties);

  const credentialCopy = Credential.removeClaimProperties(
    credential,
    needRemoving,
  );

  const collection: KiltPublishedCredentialCollectionV1 = [
    { credential: credentialCopy },
  ];

  const blob = jsonToBase64(collection);
  const url = `data:text/json;base64,${blob}`;

  return { name, url };
}

export async function invalidateCredentials(
  credentials: SporranCredential[],
): Promise<void> {
  for (const credential of credentials) {
    await saveCredential({ ...credential, status: 'invalid' });
  }
}

export async function checkCredentialsStatus(
  credentials: SporranCredential[],
): Promise<void> {
  for (const credential of credentials) {
    await syncCredentialStatusWithChain(credential);
  }
}
