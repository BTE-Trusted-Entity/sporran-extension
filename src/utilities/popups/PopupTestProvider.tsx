import { MemoryRouter } from 'react-router-dom';

import { jsonToBase64 } from '../base64/base64';

interface Props {
  path: string;
  data: unknown;
  children: JSX.Element;
}

export function PopupTestProvider({
  path,
  data,
  children,
}: Props): JSX.Element {
  const serialized = jsonToBase64(data);
  return (
    <MemoryRouter initialEntries={[`${path}?data=${serialized}`]}>
      {children}
    </MemoryRouter>
  );
}
