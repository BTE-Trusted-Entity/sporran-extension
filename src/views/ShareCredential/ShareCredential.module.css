.container {
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
  margin: 0;
}

.subline {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0.25rem 0 0.75rem 0;
}

.credentials {
  composes: table from '../../components/Table/Table.module.css';
  font-size: calc(9rem / 16);
  align-self: stretch;
  margin: 2rem 0 0.25rem;
}

.checkbox {
  composes: —Åheckbox from '../../components/Checkbox/Checkbox.module.css';
}

.checkbox + * {
  margin-right: 0.375rem;
}

.valid {
  composes: td from '../../components/Table/Table.module.css';
  background-image: url('../../components/images/checkmark.svg');
  background-repeat: no-repeat;
  background-position: center;
  background-size: 0.75rem;
}

.label {
  composes: smallText from '../../components/Typography/Typography.module.css';
  text-align: left;
  align-self: stretch;
  margin-top: 1.25rem;
}

.passwordLine {
  display: block;
  position: relative;
  margin: 0.5rem 0 1.5rem;
}

.password {
  composes: input from '../../components/Input/Input.module.css';
  width: calc(100% - 3.5rem);
  padding-right: 2.5rem;
}

.buttonsLine {
  margin: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 1rem;
}

.cancel {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}

.submit {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}
