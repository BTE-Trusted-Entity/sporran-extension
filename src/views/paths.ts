export { generatePath } from 'react-router-dom';

export const paths = {
  home: '/',
  settings: '/settings',
  access: '/access',
  identity: {
    base: '/identity/',
    add: '/identity/add',
    create: {
      start: '/identity/create',
      backup: '/identity/create/backup',
      verify: '/identity/create/verify',
      password: '/identity/create/password',
    },
    import: {
      start: '/identity/import',
      password: '/identity/import/password',
    },

    overview: '/identity/:address/:type(created|imported|pwreset)?',
    reset: {
      start: '/identity/:address/reset',
      password: '/identity/:address/reset/password',
    },
    send: {
      start: '/identity/:address/send',
      review: '/identity/:address/send/review',
      warning: '/identity/:address/send/warning',
    },
    receive: '/identity/:address/receive',
    credentials: {
      base: '/identity/:address/credentials',
      presentation: '/identity/:address/credentials/:hash/presentation',
    },
    remove: '/identity/:address/remove',
    did: {
      upgrade: {
        start: '/identity/:address/did/upgrade',
        kilt: '/identity/:address/did/upgrade/kilt',
        euro: '/identity/:address/did/upgrade/euro',
      },
      manage: {
        start: '/identity/:address/did',
        downgrade: {
          base: '/identity/:address/did/downgrade',
          warning: {
            web3name: '/identity/:address/did/downgrade/warning/web3name',
            credentials: '/identity/:address/did/downgrade/warning/credentials',
          },
          sign: '/identity/:address/did/downgrade/sign',
        },
        endpoints: {
          start: '/identity/:address/did/endpoints',
          edit: '/identity/:address/did/endpoints/:id?',
          sign: '/identity/:address/did/endpoints/sign',
        },
      },
      repair: '/identity/:address/did/repair',
    },
    web3name: {
      create: {
        base: '/identity/:address/web3name/create',
        info: '/identity/:address/web3name/create/info',
        form: '/identity/:address/web3name/create/form',
        choose: '/identity/:address/web3name/create/choose/:web3name',
        kilt: '/identity/:address/web3name/create/kilt/:web3name',
        euro: '/identity/:address/web3name/create/euro/:web3name',
      },
      manage: {
        start: '/identity/:address/web3name',
        remove: '/identity/:address/web3name/remove',
      },
    },
  },
  popup: {
    base: '/popup',
    import: '/popup/import',
    access: '/popup/access',
    claim: '/identity/:address/claim',
    save: '/popup/save',
    share: {
      start: '/popup/share',
      sign: '/popup/share/sign',
    },
    sign: '/popup/sign',
    signRaw: '/popup/signRaw',
    signDid: {
      start: '/identity/:address/did/sign',
      sign: '/identity/:address/did/sign/password',
      credentials: '/identity/:address/did/sign/credentials',
    },
    signDidExtrinsic: '/identity/:address/did/signExtrinsic',
    createDid: '/identity/:address/did/create',
  },
};
