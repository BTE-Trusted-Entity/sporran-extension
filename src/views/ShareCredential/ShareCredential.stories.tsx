import { Meta } from '@storybook/react';

import { PopupTestProvider } from '../../utilities/popups/PopupTestProvider';
import { paths } from '../paths';

import { ShareCredential } from './ShareCredential';

export default {
  title: 'Views/ShareCredential',
  component: ShareCredential,
} as Meta;

const mockClaimRequest = [
  {
    cTypeHash:
      '0xf53f460a9e96cf7ea3321ac001a89674850493e12fad28cbc868e026935436d2',
  },
];

export function Template(): JSX.Element {
  return (
    <PopupTestProvider path={paths.popup.share} data={mockClaimRequest}>
      <ShareCredential />
    </PopupTestProvider>
  );
}
