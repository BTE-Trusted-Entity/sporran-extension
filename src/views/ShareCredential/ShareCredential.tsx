import { useCallback, useState } from 'react';
import { Route, Switch } from 'react-router-dom';

import { Identity } from '../../utilities/identities/types';
import { SporranCredential } from '../../utilities/credentials/credentials';

import { shareChannel } from '../../channels/shareChannel/shareChannel';
import { paths } from '../paths';

import { ShareCredentialSign } from '../ShareCredentialSign/ShareCredentialSign';
import { ShareCredentialSelect } from '../ShareCredentialSelect/ShareCredentialSelect';
import { usePopupData } from '../../utilities/popups/usePopupData';
import { ShareInput } from '../../channels/shareChannel/types';

export interface Selected {
  sporranCredential: SporranCredential;
  identity: Identity;
  sharedContents: string[];
}

export function ShareCredential(): JSX.Element | null {
  const handleCancel = useCallback(async () => {
    await shareChannel.throw('Rejected');
    window.close();
  }, []);

  const popupData = usePopupData<ShareInput>();

  const [selected, setSelected] = useState<Selected>();

  const handleSelect = useCallback((value: Selected) => setSelected(value), []);

  return (
    <Switch>
      <Route path={paths.popup.share.sign}>
        {selected && (
          <ShareCredentialSign
            selected={selected}
            onCancel={handleCancel}
            popupData={popupData}
          />
        )}
      </Route>
      <Route path={paths.popup.share.start}>
        <ShareCredentialSelect
          onCancel={handleCancel}
          onSelect={handleSelect}
          selected={selected}
        />
      </Route>
    </Switch>
  );
}
