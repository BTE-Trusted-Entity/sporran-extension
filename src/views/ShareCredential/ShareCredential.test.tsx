import { render } from '../../testing/testing';
import { useIdentityCredentials } from '../../utilities/credentials/credentials';
import { credentialsMock } from '../../utilities/credentials/credentials.mock';
import { PopupTestProvider } from '../../utilities/popups/PopupTestProvider';
import { paths } from '../paths';

import { ShareCredential } from './ShareCredential';

jest.mock('../../utilities/credentials/credentials');
(useIdentityCredentials as jest.Mock).mockReturnValue(credentialsMock);

const mockClaimRequest = [
  {
    cTypeHash:
      '0xf53f460a9e96cf7ea3321ac001a89674850493e12fad28cbc868e026935436d2',
  },
];

describe('ShareCredential', () => {
  it('should render', async () => {
    const { container } = render(
      <PopupTestProvider path={paths.popup.share} data={mockClaimRequest}>
        <ShareCredential />
      </PopupTestProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
