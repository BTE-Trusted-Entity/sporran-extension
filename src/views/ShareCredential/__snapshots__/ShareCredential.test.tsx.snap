// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareCredential should render 1`] = `
<div>
  <form
    class="container"
  >
    <h1
      class="heading"
    >
      Your Credentials
    </h1>
    <p
      class="subline"
    >
      Pick credentials to sign and share
    </p>
    <section>
      <div
        class="MacFarlane tartan"
      >
        Identicon
      </div>
      <div
        class="nameLine"
      >
        <span
          class="name"
        >
          My Sporran Account
        </span>
        <div
          class="wrapper"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            aria-label="Account options"
            class="toggle"
            role="button"
            tabindex="0"
            title="Account options"
            type="button"
          />
        </div>
      </div>
    </section>
    <table
      class="credentials"
    >
      <thead>
        <tr
          class="tr"
        >
          <th
            class="th"
          >
            <label>
              <input
                class="checkbox"
                name="all"
                type="checkbox"
              />
              <span />
              all
            </label>
          </th>
          <th
            class="th"
          >
            Name
          </th>
          <th
            class="th"
          >
            C-Type
          </th>
          <th
            class="th"
          >
            Attester
          </th>
          <th
            class="th"
          >
            Valid
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="tr"
        >
          <td
            class="td"
          >
            <label>
              <input
                aria-label="Share credential"
                checked=""
                class="checkbox"
                name="0"
                type="checkbox"
              />
              <span />
            </label>
          </td>
          <td
            class="td"
          >
            email private
          </td>
          <td
            class="td"
          >
            BL-Mail-Simple
          </td>
          <td
            class="td"
          >
            socialKYC
          </td>
          <td
            aria-label="Valid"
            class="valid"
          />
        </tr>
        <tr
          class="tr"
        >
          <td
            class="td"
          >
            <label>
              <input
                aria-label="Share credential"
                class="checkbox"
                name="1"
                type="checkbox"
              />
              <span />
            </label>
          </td>
          <td
            class="td"
          >
            twitter
          </td>
          <td
            class="td"
          >
            BL-Mail-Advanced
          </td>
          <td
            class="td"
          >
            socialKYC
          </td>
          <td
            aria-label="Valid"
            class="valid"
          />
        </tr>
      </tbody>
    </table>
    <label
      class="label"
    >
      Sign with Password:
      <span
        class="passwordLine"
      >
        <input
          class="password"
          name="password"
          required=""
          type="password"
        />
        <button
          aria-label="Show"
          class="show"
          title="Show"
          type="button"
        />
      </span>
    </label>
    <p
      class="buttonsLine"
    >
      <button
        class="cancel"
        type="button"
      >
        Cancel
      </button>
      <button
        class="submit"
        disabled=""
        type="submit"
      >
        Share
      </button>
    </p>
  </form>
</div>
`;
